lib/services/local/notifications.go

```diff
@@ -0,0 +1,298 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = serviceWithPrefix.DeleteAllResources(ctx)
+	return trace.Wrap(err)
+}
+
+// ListUserNotificationStates returns a page of a user's notification states.
+func (s *NotificationsService) ListUserNotificationStates(ctx context.Context, username string, pageSize int, nextToken string) ([]*notificationsv1.UserNotificationState, string, error) {
+	// Append username prefix.
+	serviceWithPrefix, err := s.userNotificationStateService.WithPrefix(username)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	states, nextToken, err := serviceWithPrefix.ListResources(ctx, pageSize, nextToken)
+	return states, nextToken, trace.Wrap(err)
+}
+
+// UpsertUserLastSeenNotification creates or updates a user's last seen notification item.
+func (s *NotificationsService) UpsertUserLastSeenNotification(ctx context.Context, username string, ulsn *notificationsv1.UserLastSeenNotification) (*notificationsv1.UserLastSeenNotification, error) {
+	if err := services.ValidateUserLastSeenNotification(ulsn); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ulsn.Kind = types.KindUserLastSeenNotification
+	ulsn.Version = types.V1
+
+	if ulsn.Metadata == nil {
+		ulsn.Metadata = &headerv1.Metadata{}
+	}
+	// We set this to the username because the service adapter uses `getName()` to determine the backend key to use when storing the object.
+	ulsn.Metadata.Name = username
+
+	upserted, err := s.userLastSeenNotificationService.UpsertResource(ctx, ulsn)
+	return upserted, trace.Wrap(err)
+}
+
+// GetUserLastSeenNotification returns a user's last seen notification item.
+func (s *NotificationsService) GetUserLastSeenNotification(ctx context.Context, username string) (*notificationsv1.UserLastSeenNotification, error) {
+
+	ulsn, err := s.userLastSeenNotificationService.GetResource(ctx, username)
+	return ulsn, trace.Wrap(err)
+}
+
+// DeleteUserLastSeenNotification deletes a user's last seen notification item.
+func (s *NotificationsService) DeleteUserLastSeenNotification(ctx context.Context, username string) error {
+	err := s.userLastSeenNotificationService.DeleteResource(ctx, username)
+	return trace.Wrap(err)
+}
+
+// CheckAndSetExpiry checks and sets the default expiry for a notification.
+func CheckAndSetExpiry(expires **timestamppb.Timestamp) error {
+	// If the expiry hasn't been provided, set the default to 30 days from now.
+	if *expires == nil {
+		now := time.Now()
+		futureTime := now.Add(30 * 24 * time.Hour)
+		*expires = timestamppb.New(futureTime)
+	}
+
+	// If the expiry has already been provided, ensure that it is not more than 90 days from now.
+	// This is to prevent misuse as we don't want notifications existing for too long and accumulating in the backend.
+	now := time.Now()
+	maxExpiry := now.Add(90 * 24 * time.Hour)
+
+	if (*expires).AsTime().After(maxExpiry) {
+		return trace.BadParameter("notification expiry cannot be more than 90 days from its creation")
+	}
```

Comment:
Wouldn't it be bad practice to quietly overwrite a client-configured field? I think the error is useful to prevent misuse, otherwise a client could be trying to intentionally configure an expiry of 91+ days oblivious to the fact it's invalid and actually being overwritten.